### ISSUE TITLE
Report whether an agent is Fleet managed or standalone or unprivileged to components

### ISSUE THREAD

**Issue by cmacknz** [elastic/elastic-agent#4683]
- Relates https://github.com/elastic/elastic-agent/issues/3065
- Relates https://github.com/elastic/beats/pull/39403#issuecomment-2094754790

Allow components (specifically Beats sub-processes) to determine if they are connected to a standalone or a Fleet managed agent so that they can add this information to the User-Agent string. Additionally include whether the agent is unprivileged since it is cheap to add as part of this change.

Determining the privilege level is a single function call, for example see https://github.com/elastic/elastic-agent/blob/eaf3c4343603e818b5acb5ad1364a718551d9b8c/pkg/control/v2/server/server.go#L401

Adding this information to the `AgentInfo` message in the control protocol is likely the best choice. https://github.com/elastic/elastic-agent-client/blob/ac72c3488d26b2b319cec87536d45eea35b81d17/elastic-agent-client.proto#L197-L210

```protobuf
// Agent information that the component might want to use for its events,
// including the package version, which components should report instead
// of their own version.
//
// Sent on component start up as part of StartUpInfo and on the first checkin
// expected response to the component.
message AgentInfo {
    // ID is the Elastic Agent's unique ID.
    string id = 1;
    // Version is the package version of the running Elastic Agent.
    string version = 2;
    // Snapshot is true when the running Elastic Agent is a snapshot version.
    bool snapshot = 3;
}
```

**Comment by pierrehilbert** [comment]
@fearful-symmetry mentioned to me it was less trivial than what we initially thought to get the unprivileged info in the User Agent.
@amitkanfer as you were the one pushing to get this one in the User Agent, what value are we trying to get from it as we already have the telemetry for this feature usage? I just want to check if we need to continue to investigate time on this or if it was just "a nice to have" thing.

**Comment by amitkanfer** [comment]
>  what value are we trying to get from it as we already have the telemetry for this feature usage

What feature? unprivileged or fleet managed? If we have telemetry for everything so we should probably close https://github.com/elastic/elastic-agent/issues/3065 and https://github.com/elastic/beats/pull/39403#issuecomment-2094754790 as well.

> mentioned to me it was less trivial than what we initially thought to get the unprivileged info in the User Agent.

@pierrehilbert can you please describe why is this not trivial? that's what bothers me the most tbh

**Comment by pierrehilbert** [comment]
> What feature? unprivileged or fleet managed? If we have telemetry for everything so we should probably close #3065 and [elastic/beats#39403 (comment)](https://github.com/elastic/beats/pull/39403#issuecomment-2094754790) as well.

Unprivileged: the only case we are not currently covering is standalone Agent running as unprivileged (this option is not available for Beats standalone). I would like to understand how important it is to get this information to decide if it should remain a priority to spend time on.

> @pierrehilbert can you please describe why is this not trivial? that's what bothers me the most tbh

I will let @fearful-symmetry reply to this question.

**Comment by fearful-symmetry** [comment]
Regarding the time these changes take:

The process of getting data (like management mode state) from agent's state to beat's `User-Agent` HTTP body required changes to the gRPC protocol, which required changes to three repos: `elastic-agent`, `elastic-agent-client` and `beats`. This adds a lot of overhead to process, as we now have three PRs, with two or three separate tests to write, that need code reviews, that need to pass CI, that require familiarity with three different codebases, etc. In the past this isn't much of a problem, as we haven't needed to make frequent changes to the client gRPC API. However, for something seemingly simple, like passing the management mode to the beat's user-agent, this requires a disproportionate amount of time. 

My understanding is that we want all this data added to user-agent for the sake of telemetry; in general, the system we have now, (a set gRPC protocol that will have to move metadata from one component to another, a  legacy elasticsearch client in beats that we must adapt in order to add to the user-agent, etc) was not designed with telemetry in mind. I think it may be best if we design some kind of dedicated telemetry component that can more elegantly pass state information to ES, and (more importantly) can be changed and added to with relative ease, as opposed to piecemeal changes to every part of the stack. This would probably be a good use of our time, as (in my experience) the things that BI/telemetry cares about tend to change. This component can either act independently of the beats ES output, or we can refactor the output to more elegantly accept config from a generic telemetry component. 

Regarding the privilege mode reporting, it is possible to do this relatively simply: we just report if beats is running as `root` (or whatever the windows equivalent is). This would be a fairly small change. However, I'm under the assumption that we don't want to do that, and instead report the actual config state: has agent itself been configured _by the user to run in unprivileged mode?_ Due to bugs or environment config, there can be a discrepancy between what the user configured versus what's _actually_ running, and I assume we care about the former, not the latter. However, that's more of a philosophical question about what the metric represents. If we decide we only care about what user beats is running as, independent from how the user may have configured agent, that change could be done in 1-2 days.

**Comment by amitkanfer** [comment]
Thank you for the detailed response @fearful-symmetry , makes sense. @nimarezainia, for unprivileged telemetry, are we interested in what the user configured the agent to run in, or the actual running mode? (as Alex wrote, could be different in some bad cases).

**Comment by nimarezainia** [comment]
@amitkanfer the main purpose of reporting the user-agent was so that we can better determine the volume of data ingress to ESS, more importantly differentiate between agent and standalone beats generated data. It may be important imo to know whether the **volume** is attributed to priv or unpriv, but I don't have an immediate use case for this. To be able to respond to Gartner we wouldn't need this fidelity. 

We have other telemetry that will show indicatively whether this feature is being used. In product  knowing if the feature is being used and by what % of deployments is our priority.

I'm a bit confused however with how a beat that's spawned up by agent can be in a different mode. Maybe a transient state or a bug but in general If Agent is configured as `unprivileged` we should assume that the beats spawned by it will also be in the same mode ?  if this is not the case we have a bug. I'm not sure reporting mode via user-agent is intended to catch these bugs.

If we go down this path I would say we need to know what the Agent was configured to run as, as in what is the user's intention. but perhaps the mode the beats is running in can be used as the proxy for what the agent is configured to run as.

**Comment by fearful-symmetry** [comment]
Yeah, the more I think about it, the more I think we should report configuration as apposed to user state; over the years I've seen at least a handful of users get creative and attempt to run agent as non-root.
